### ISSUE TITLE
Config dependency manager doesn't take into account hooks until "afterResolvers".

### ISSUE THREAD

**Issue by reves** [webpack/webpack#16312]
# Bug report

**What is the current behavior?**
I use [configuration dependencies](https://webpack.js.org/configuration/configuration-types/#dependencies) to make sure that the plugins from the second config `'server'` will always be executed after the first config `'client'` output is emitted:

```javascript
module.exports = [
  {
    name: 'client',
    plugins: [
      {
        apply: compiler => compiler.hooks.afterEmit.tap('Plugin1', _ => {
          console.log('Config 1: afterEmit')
        })
      }
    ]
  },
  {
    name: 'server',
    dependencies: ['client'],
    plugins: [
      {
        apply: compiler => compiler.hooks.afterResolvers.tap('Plugin2', _ => {
          // I picked the "afterResolvers" hook, because i will apply here the HtmlWebpackPlugin,
          // that is required to be applied before the "initialize" step.
          console.log('Config 2: afterResolvers')
        })
      }
    ]
  }
]
```
Doesn't work for [hooks](https://webpack.js.org/api/compiler-hooks/#hooks) like `afterResolvers` (or `initialize`):
```
Config 2: afterResolvers
Config 1: afterEmit
```
However, this works as expected starting from the `beforeRun` hook:
```
Config 1: afterEmit
Config 2: beforeRun
```

**What is the expected behavior?**

```
Config 1: afterEmit
Config 2: afterResolvers
```

**Other relevant information:**
webpack version:  5.53.0
Node.js version: 16.17.0
Operating System: Windows 10

**Comment by alexander-akait** [comment]
I don't think it is bug, I want to say it is hidden feature, because https://github.com/webpack/webpack/blob/main/lib/WebpackOptionsApply.js#L695, i.e. `afterResolvers` is configuration hook, and all configuration hooks runs concurency

You can add more deps or change the whole configuration in `afterResolvers`, so we need do it before runs

**Comment by alexander-akait** [comment]
`HtmlWebpackPlugin` should not use `compiler.hooks.initialize` to get options, that is wrong and invalid, options can be change, so it need to get in the valid hooks, but here workaround:

```js
const Plugin = require("html-webpack-plugin")

const compiler = [
  {
    name: 'client',
    output: {
      filename: "client-[name].js"
    },
    plugins: [
      {
        apply: compiler => compiler.hooks.afterEmit.tapAsync('Plugin1', (compilation, callback) => {
          console.log('Config 1: afterEmit')

          callback()
        })
      }
    ]
  },
  {
    name: 'server',
    dependencies: ['client'],
    output: {
      filename: "server-[name].js"
    },
    plugins: [
      {
        apply: compiler => compiler.hooks.beforeRun.tap('Plugin2', (compiler) => {
          new Plugin().apply(compiler);

          compiler.hooks.initialize.call(compiler);
        })
      }
    ]
  }
]

module.exports = compiler;
```

/cc @jantimon can we rewrite it, to get entries you can use this hook - https://github.com/webpack/webpack/blob/00c57ede11ee10dbb0a42e944b2cd7a038089944/lib/ProgressPlugin.js#L393, other things is just default options you don't need it to do inside hook and it is wrong

**Comment by reves** [comment]
> all configuration hooks runs concurency

I guessed about it, but I wasn't sure. Thanks for pointing this out.


> but here workaround:
[...]
compiler.hooks.initialize.call(compiler)

Such a simple and smart solution. It works great, thank you!

**What's the point?**
I execute the emitted script (it returns static HTML for every route of the client app) and then i apply the `HtmlWebpackPlugin` to generate `.html` files.
